### PR TITLE
Feature: Introducing embedded Trace Viewer

### PR DESCRIPTION
**Pull request by ruifigueira**
See https://github.com/microsoft/playwright-vscode/pull/483#pullrequestreview-2195415779

### CODE REVIEW

File: package.json
Comment (pavelfeldman) [inline package.json:73]: Let's keep it as a hidden setting at first, not visible in a sidebar, no toggle action.

File: package.json
Comment (pavelfeldman) [inline package.json:122]: I.e. we will only have this.

File: src/settingsView.ts
Comment (pavelfeldman) [inline src/settingsView.ts:281]: Remove this

File: src/embeddedTraceViewer.ts
Comment (pavelfeldman) [inline src/embeddedTraceViewer.ts:34]: I don't think embedded trace viewer makes sense w/o the test server, make it required?

File: src/embeddedTraceViewer.ts
Comment (pavelfeldman) [inline src/embeddedTraceViewer.ts:101]: This code has inherent race condition where calling it 5 times will create 5 panel instances.

File: src/embeddedTraceViewer.ts
Comment (pavelfeldman) [inline src/embeddedTraceViewer.ts:170]: What else are we expecting? Is it safe to destructure it one line above?

File: src/playwrightTestServer.ts
Comment (pavelfeldman) [inline src/playwrightTestServer.ts:47]: It is unclear this starts the server. Let's make it more clear and specific: ensureStartedForTraceViewer(): and return trace viewer URL

File: src/testModel.ts
Comment (pavelfeldman) [inline src/testModel.ts:112]: It does not look like this is necessarily a user gesture. Looks like a generic programmatic getter to me

File: src/testModel.ts
Comment (pavelfeldman) [inline src/testModel.ts:395]: The name does not make sense to me - I think once you closed the trace viewer you can't use it. So either traceViewer should be a traceViewerFactory, or this should be closeTraceView().

File: src/traceViewer.ts
Comment (pavelfeldman) [inline src/traceViewer.ts:47]: nit: Passing settingsModel does not seem justified atm, but I'm fine with it

File: tests-integration/tests/baseTest.ts
Comment (pavelfeldman) [inline tests-integration/tests/baseTest.ts:83]: I'm not reviewing this as a part of this change, could you please extract it and explain what and why is happening here?

File: tests/utils.ts
Comment (pavelfeldman) [inline tests/utils.ts:37]: Let's land the testing harness changes separately, before this change.

**Review by pavelfeldman — COMMENTED** [review]
Looks great, thanks!

**Comment by pavelfeldman** [comment]
I also noticed a bunch of bugs, but we don't need to fix them right now. Selection seems to be broken and dragging the view between tabs does not restore its state (this one will be interesting)

File: src/embeddedTraceViewer.ts
Comment (ruifigueira) [reply src/embeddedTraceViewer.ts:34]: I was assuming that the **Embedded** checkbox would be available even in legacy versions that had `PlaywrightTestCLI`, and that way the `EmbeddedTraceViewer` instance would show the warning message.

But if the warning message is not needed, I can make it required and only instanciate `EmbeddedTraceViewer` when it makes sense. In that case, I can also move the `checkSupport` logic into `testModel` and remove `settingsModel` dependency from trace viewer.

File: src/traceViewer.ts
Comment (ruifigueira) [reply src/traceViewer.ts:47]: I think changes I'll do on https://github.com/microsoft/playwright-vscode/pull/513#discussion_r1706132388 will no longer require this dependency on `settingsModel`

File: tests-integration/tests/baseTest.ts
Comment (ruifigueira) [reply tests-integration/tests/baseTest.ts:83]: I'll remove the integration test commit from this PR, and I'll include it after these merges are done.

File: tests/utils.ts
Comment (ruifigueira) [reply tests/utils.ts:37]: done: https://github.com/microsoft/playwright-vscode/pull/514

File: src/testModel.ts
Comment (ruifigueira) [reply src/testModel.ts:395]: I was keeping the same pattern from `TraceViewer` before my changes: one instance with the same lifecycle as its container (before it was `extension`, now it's `testModel`) that could be closed and open several times.

And the idea of this `_closeTraceViewerIfNeeded` is to only close the trace viewer when either the selected model actually changes or when current trace viewer becomes due to settings changes (e.g., checking the embedded setting would close the spawn trace viewer and open the embedded),

There are situations where changing trace related settings won't close the current trace viewer. For instance, if it currently has embedded checked but playwright version doesn't support it, it will open spawn trace viewer. If the user then unchecks the embedded setting, it should keep the spawn trace viewer open.

File: src/embeddedTraceViewer.ts
Comment (ruifigueira) [reply src/embeddedTraceViewer.ts:170]: Messages that get here are from `playwright/packages/trace-viewer` module.

I was reviewing the changes I did there and unfortunatelly I used different message keys, both `command` and `message`:

- `command: 'openExternal'`: 

https://github.com/microsoft/playwright/blob/7ec3a93db3607e39b001e700020635bb017b4508/packages/trace-viewer/src/ui/embeddedWorkbenchLoader.tsx#L26

- `method: 'showErrorMessage'`:

https://github.com/microsoft/playwright/blob/7ec3a93db3607e39b001e700020635bb017b4508/packages/trace-viewer/src/ui/embeddedWorkbenchLoader.tsx#L82

So I need to change this, and I'll also create a new PR in playwright to make them consistent (with `method`, which is already very common in playwright code for these kind of messages).

**Comment by ruifigueira** [comment]
I'm also experiencing some odd behaviour when launching tests in debug. From what I can tell, when debugging from vscode, it deactivates tracing:

https://github.com/microsoft/playwright-vscode/blob/e94a82316ca54e45f27025e5f44908a344b8f1a4/src/testModel.ts#L397

but nevertheless, it sets the live json path in the test item:

https://github.com/microsoft/playwright-vscode/blob/98c19d608e0983b8086de42dace7fff53c626b10/src/extension.ts#L461-L462

This is causing the trace viewer to open in debug, without displaying any trace because the path is not valid.

**Comment by ruifigueira** [comment]
> dragging the view between tabs does not restore its state

This option does the trick: https://code.visualstudio.com/api/extension-guides/webview#retaincontextwhenhidden .

It doesn't work when moving the panel into a new window, though.

File: src/embeddedTraceViewer.ts
Comment (pavelfeldman) [reply src/embeddedTraceViewer.ts:34]: Yes, let's only support Embedded for modern PlaywrightTestServer-based configurations only.

File: src/testModel.ts
Comment (pavelfeldman) [reply src/testModel.ts:395]: Let's land this bit as is and if I come up with a better name I'll follow up.

**Comment by pavelfeldman** [comment]
> This is causing the trace viewer to open in debug, without displaying any trace because the path is not valid.

If this bug exists on tip of the tree we should file and fix it

**Comment by ruifigueira** [comment]
I already open a PR to fix that bug: https://github.com/microsoft/playwright-vscode/pull/515

File: src/embeddedTraceViewer.ts
Comment (ruifigueira) [reply src/embeddedTraceViewer.ts]: Addressed in 31a15f3f926c5f17a8e16a4476d52c62f11778a9

File: src/embeddedTraceViewer.ts
Comment (ruifigueira) [reply src/embeddedTraceViewer.ts]: Addressed in 7b384dbe30b8ccc1fde6d310adb668c8a9421e65

File: src/playwrightTestServer.ts
Comment (ruifigueira) [reply src/playwrightTestServer.ts]: Addressed in 49d717f955696790e6d61abb7e91cf29ba63afc1

File: src/testModel.ts
Comment (ruifigueira) [reply src/testModel.ts]: Addressed in 31a15f3f926c5f17a8e16a4476d52c62f11778a9 with `traceViewer()` and `ensureTraceViewer()`. The latter creates a trace viewer if not initialized, as a result of a user gesture (running tests or selecting a test item).

File: src/testModel.ts
Comment (ruifigueira) [reply src/testModel.ts]: I left the `_closeTraceViewerIfNeeded` function but created a traceViewerFactory in 31a15f3f926c5f17a8e16a4476d52c62f11778a9 to check which trace viewer strategy to instanciate.

File: src/traceViewer.ts
Comment (ruifigueira) [reply src/traceViewer.ts]: Addressed in 31a15f3f926c5f17a8e16a4476d52c62f11778a9

File: src/testModel.ts
Comment (pavelfeldman) [inline src/testModel.ts:61]: Move this into its own file?

File: src/traceViewer.ts
Comment (pavelfeldman) [inline src/traceViewer.ts:35]: Move this out into its own file as a separate change and then you can place TraceViewerFactory here.

File: src/embeddedTraceViewer.ts
Comment (ruifigueira) [reply src/embeddedTraceViewer.ts:69]: It was my initial thought, but I couldn't dispose `traceViewerPanel` synchronously and forces `infoForTest` to become async. But I'll do it that way

Comment (ruifigueira) [reply src/embeddedTraceViewer.ts]: Addressed in 1942a463a88da42e1ba8a1d4628c57476eaa4ff6

File: src/embeddedTraceViewer.ts
Comment (ruifigueira) [reply src/embeddedTraceViewer.ts]: Addressed in 1942a463a88da42e1ba8a1d4628c57476eaa4ff6

File: src/traceViewer.ts
Comment (ruifigueira) [reply src/traceViewer.ts]: Addressed in 87543434f44c7494c7e7e46678f88bd4173db83a

File: package.nls.json
Comment (pavelfeldman) [inline package.nls.json:17]:
```suggestion
  "configuration.playwright.embeddedTraceViewer": "Toggle Embedded Playwright Trace Viewer",
```

File: l10n/bundle.l10n.json
Comment (pavelfeldman) [inline l10n/bundle.l10n.json:21]: Unused, revert

File: src/testModel.ts
Comment (pavelfeldman) [inline src/testModel.ts:121]: This will never close spawned trace viewer since it is always supported.

File: src/testModel.ts
Comment (pavelfeldman) [inline src/testModel.ts:103]: For the embedded version, I expect this toggle to show the trace viewer upon enabling.

File: src/embeddedTraceViewer.ts
Comment (pavelfeldman) [inline src/embeddedTraceViewer.ts:107]: Since we are not implementing an editor of the selected document, we should not use this API to show trace viewer. We should instead show it as a sidebar or a drawer panel and encourage user to drag it to the right. Try doing it with the Playwright settings and you'll see what I mean.

**Review by pavelfeldman — APPROVED** [review]
The change looks great. It still has several small problems, but I am open to landing it as is as long as you'll follow up to these issues!

File: l10n/bundle.l10n.json
Comment (ruifigueira) [reply l10n/bundle.l10n.json]: Addressed in b8a3d8b105bfd17b1c4c1dd8b4263723356ef219

File: package.nls.json
Comment (ruifigueira) [reply package.nls.json]: Addressed in b8a3d8b105bfd17b1c4c1dd8b4263723356ef219

File: src/testModel.ts
Comment (ruifigueira) [reply src/testModel.ts]: Addressed in eaf9074019931773ad149d158711f9184846c4db

File: src/embeddedTraceViewer.ts
Comment (ruifigueira) [reply src/embeddedTraceViewer.ts:112]: Do you mean this?

image

I honestly don't really like that approach. I'd rather go with editors. I already started to implement it but I left it on hold while this PR was under review.

For reference, here's a screenshot of the current status, with two trace editors side by side:

image

I did use the `vscode.window.createWebviewPanel` API because it's the one that offers a similar user experience to editors.

File: l10n/bundle.l10n.json
Comment (pavelfeldman) [inline l10n/bundle.l10n.json:26]: why touching this file?

File: src/testModel.ts
Comment (pavelfeldman) [reply src/testModel.ts:124]: I was referring to changing the settings value. If you change the setting value nothing will happen.

File: src/embeddedTraceViewer.ts
Comment (pavelfeldman) [reply src/embeddedTraceViewer.ts:112]: Traces side by side is a good example of why you should not be using the editor pattern. It is unclear what controls what. If trace viewer was opened for a trace file, it would be a different story.

**Comment by pavelfeldman** [comment]
I landed the change because every time you change one line I need to review this whole huge PR, it is time consuming. Please make incremental changes that can be code-reviewed, otherwise we are at risk of compromising the product quality!

File: src/embeddedTraceViewer.ts
Comment (ruifigueira) [reply src/embeddedTraceViewer.ts:112]: I created URIs so that they got proper breadcrumbs.

Meanwhle, I was able to implement it with webviews. Here's a screenshot of it:

image

, I'll clean the code a bit and create a new PR for that when I have some spare time.

**Pull request merged**
